userspace/falco/grpc_request_context.cpp

```diff
@@ -0,0 +1,113 @@
+limitations under the License.
+*/
+
+#include "grpc_request_context.h"
+
+namespace falco
+{
+namespace grpc
+{
+
+template<>
+void request_stream_context<falco::output::service, falco::output::request, falco::output::response>::start(server* srv)
+{
+	m_state = request_context_base::REQUEST;
+	m_srv_ctx.reset(new ::grpc::ServerContext);
+	auto srvctx = m_srv_ctx.get();
+	m_res_writer.reset(new ::grpc::ServerAsyncWriter<output::response>(srvctx));
+	m_stream_ctx.reset();
+	m_req.Clear();
+	auto cq = srv->m_completion_queue.get();
+	(srv->m_output_svc.*m_request_func)(srvctx, &m_req, m_res_writer.get(), cq, cq, this);
+}
+
+template<>
+void request_stream_context<falco::output::service, falco::output::request, falco::output::response>::process(server* srv)
+{
+	// When it is the 1st process call
+	if(m_state == request_context_base::REQUEST)
+	{
+		m_state = request_context_base::WRITE;
+		m_stream_ctx.reset(new stream_context(m_srv_ctx.get()));
+	}
+
+	// Processing
+	output::response res;
+	(srv->*m_process_func)(*m_stream_ctx, m_req, res); // subscribe()
+
+	// When there still are more responses to stream
+	if(m_stream_ctx->m_has_more)
+	{
+		m_res_writer->Write(res, this);
+	}
+	// No more responses to stream
+	else
+	{
+		// Communicate to the gRPC runtime that we have finished.
+		// The memory address of `this` instance uniquely identifies the event.
+		m_state = request_context_base::FINISH;
+		m_res_writer->Finish(::grpc::Status::OK, this);
+	}
+}
+
+template<>
+void request_stream_context<falco::output::service, falco::output::request, falco::output::response>::end(server* srv, bool errored)
+{
+	if(m_stream_ctx)
+	{
+		m_stream_ctx->m_status = errored ? stream_context::ERROR : stream_context::SUCCESS;
+
+		// Complete the processing
+		output::response res;
+		(srv->*m_process_func)(*m_stream_ctx, m_req, res); // subscribe()
+	}
+
+	start(srv);
+}
+
+template<>
+void falco::grpc::request_context<falco::version::service, falco::version::request, falco::version::response>::start(server* srv)
+{
+	m_state = request_context_base::REQUEST;
+	m_srv_ctx.reset(new ::grpc::ServerContext);
+	auto srvctx = m_srv_ctx.get();
+	m_res_writer.reset(new ::grpc::ServerAsyncResponseWriter<version::response>(srvctx));
+	m_req.Clear();
+	auto cq = srv->m_completion_queue.get();
+	(srv->m_version_svc.*m_request_func)(srvctx, &m_req, m_res_writer.get(), cq, cq, this);
+}
+
+template<>
+void falco::grpc::request_context<falco::version::service, falco::version::request, falco::version::response>::process(server* srv)
+{
+	version::response res;
+	(srv->*m_process_func)(m_srv_ctx.get(), m_req, res); // version()
+	// Done
+	m_state = request_context_base::FINISH;
+	m_res_writer->Finish(res, ::grpc::Status::OK, this);
+}
+
+template<>
+void falco::grpc::request_context<falco::version::service, falco::version::request, falco::version::response>::end(server* srv, bool errored)
+{
+	// todo(leodido) > what to do when errored is true?
```

Comment:
Should we include managing this error here? Would this potentially crash the program? At the very least log and "die nice"

Reply:
This can't crash the program, it just asks the gRPC engine to start again listening for events.
Anyway, I've inserted TODOs about logging the error (to assess in future PRs)